src/lib/messaging.ts

```diff
@@ -1,135 +1,84 @@
-    | "browser_proxy_background"
-    | "download_background"
-    | "performance_background"
-export type MessageType = TabMessageType | NonTabMessageType
-
-export interface Message {
-    type: MessageType
-    // and other unknown attributes...
-    [key: string]: any
+interface Message<Root, Type extends keyof Root, Command extends keyof Root[Type]> {
+    type: Type
+    command: Command
+    args: Parameters<Root[Type][Command]>
 }
 
-export type listener = (
-    message: Message,
-    sender?,
-    sendResponse?,
-) => void | Promise<any>
-
-// Calls methods on obj that match .command and sends responses back
-export function attributeCaller(obj) {
-    function handler(message: Message, sender, sendResponse) {
-        logger.debug(message)
-
-        // Args may be undefined, but you can't spread undefined...
-        if (message.args === undefined) message.args = []
-
-        // Call command on obj
-        try {
-            const response = obj[message.command](...message.args)
+function backgroundHandler<
+    Root,
+    Type extends keyof Root,
+    Command extends (keyof Root[Type])
+    >(root: Root,
+      message: Message<Root, Type, Command>,
+      sender: browser.runtime.MessageSender,
+    ): ReturnType<Root[Type][Command]> {
+    return root[message.type][message.command](...message.args)
+}
 
-            // Return response to sender
-            if (response instanceof Promise) {
-                logger.debug("Returning promise...", response)
-                sendResponse(response)
-                // Docs say you should be able to return a promise, but that
-                // doesn't work.
-                /* return response */
-            } else if (response !== undefined) {
-                logger.debug("Returning synchronously...", response)
-                sendResponse(response)
-            }
-        } catch (e) {
-            logger.error(
-                `Error processing ${message.command}(${message.args})`,
-                e,
-            )
-            return Promise.reject(e)
+export function setupListener<Root>(root: Root) {
+    browser.runtime.onMessage.addListener((message: any, sender: browser.runtime.MessageSender) => {
+        if (message.type in root) {
+            if (!(message.command in root[message.type]))
+                throw new Error(`missing handler in protocol ${message.type} ${message.command}`)
+            if (!Array.isArray(message.args))
+                throw new Error(`wrong arguments in protocol ${message.type} ${message.command}`)
+            return backgroundHandler(root, message, sender)
         }
-    }
-    return handler
+    });
 }
 
 /** Send a message to non-content scripts */
-export async function message(type: NonTabMessageType, command, args?) {
-    return browser.runtime.sendMessage({ type, command, args } as Message)
+export function message<Root>() {
+    return async <Type extends keyof Root, Command extends keyof Root[Type]>(
```

Comment:
returning this closure sucks, but I was unable to come up with a better solution.

When I tried default values for some parameters, to be able to invoke `message<Messages.Background>(...)`, it would never infer types properly. The other thing I tried was `message(Messages.Background, ...)`, but this quickly backfires because it's used as a value and the module needs to be imported at runtime, which breaks all kinds of things if we import code across background & content scripts.